In [1]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns


sns.set()

#-------------------- 차트 관련 속성 (한글처리, 그리드) -----------
plt.rcParams['font.family']= 'Malgun Gothic'
plt.rcParams['axes.unicode_minus'] = False

#-------------------- 주피터 , 출력결과 넓이 늘리기 ---------------
# from IPython.core.display import display, HTML
from IPython.display import display, HTML

display(HTML("<style>.container{width:100% !important;}</style>"))
pd.set_option('display.max_rows', 100)
pd.set_option('display.max_columns', 100)
pd.set_option('max_colwidth', None)

import warnings
warnings.filterwarnings(action='ignore')

In [2]:
# import pandas_profiling

# ----------------- 학습 ----------------------
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler, MinMaxScaler, RobustScaler

# ----------------- 모델 ----------------------
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import Ridge

# ----------------- 평가 ----------------------
from sklearn.metrics import confusion_matrix
from sklearn.metrics import mean_squared_error
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score, precision_score, recall_score
# from sklearn.metrics import precision_recall_curve, plot_precision_recall_curve

# from sklearn.metrics import roc_auc_score, roc_curve, plot_roc_curve

# 데이터 로드

In [3]:
train = pd.read_csv('../data/traindf.csv')
test = pd.read_csv('../data/testdf.csv')
train = train.rename(columns={'Unnamed: 0':'날짜'})
test = test.rename(columns={'Unnamed: 0':'날짜'})

# train.info(),test.info()

# h2o

<pre>
import h2o
from h2o.automl import H2OAutoML

# Start the H2O cluster (locally)
h2o.init()

# Import a sample binary outcome train/test set into H2O
train = h2o.import_file("https://s3.amazonaws.com/erin-data/higgs/higgs_train_10k.csv")
test = h2o.import_file("https://s3.amazonaws.com/erin-data/higgs/higgs_test_5k.csv")

# Identify predictors and response
x = train.columns
y = "response"
x.remove(y)

# For binary classification, response should be a factor
train[y] = train[y].asfactor()
test[y] = test[y].asfactor()

# Run AutoML for 20 base models
aml = H2OAutoML(max_models=20, seed=1)
aml.train(x=x, y=y, training_frame=train)

# View the AutoML Leaderboard
lb = aml.leaderboard
lb.head(rows=lb.nrows)  # Print all rows instead of default (10 rows)

# model_id                                                  auc    logloss    mean_per_class_error      rmse       mse
# ---------------------------------------------------  --------  ---------  ----------------------  --------  --------
# StackedEnsemble_AllModels_AutoML_20181212_105540     0.789801   0.551109                0.333174  0.43211   0.186719
# StackedEnsemble_BestOfFamily_AutoML_20181212_105540  0.788425   0.552145                0.323192  0.432625  0.187165
# XGBoost_1_AutoML_20181212_105540                     0.784651   0.55753                 0.325471  0.434949  0.189181
# XGBoost_grid_1_AutoML_20181212_105540_model_4        0.783523   0.557854                0.318819  0.435249  0.189441
# XGBoost_grid_1_AutoML_20181212_105540_model_3        0.783004   0.559613                0.325081  0.435708  0.189841
# XGBoost_2_AutoML_20181212_105540                     0.78136    0.55888                 0.347074  0.435907  0.190015
# XGBoost_3_AutoML_20181212_105540                     0.780847   0.559589                0.330739  0.43613   0.190209
# GBM_5_AutoML_20181212_105540                         0.780837   0.559903                0.340848  0.436191  0.190263
# GBM_2_AutoML_20181212_105540                         0.780036   0.559806                0.339926  0.436415  0.190458
# GBM_1_AutoML_20181212_105540                         0.779827   0.560857                0.335096  0.436616  0.190633
# GBM_3_AutoML_20181212_105540                         0.778669   0.56179                 0.325538  0.437189  0.191134
# XGBoost_grid_1_AutoML_20181212_105540_model_2        0.774411   0.575017                0.322811  0.4427    0.195984
# GBM_4_AutoML_20181212_105540                         0.771426   0.569712                0.33742   0.44107   0.194543
# GBM_grid_1_AutoML_20181212_105540_model_1            0.769752   0.572583                0.344331  0.442452  0.195764
# GBM_grid_1_AutoML_20181212_105540_model_2            0.754366   0.918567                0.355855  0.496638  0.246649
# DRF_1_AutoML_20181212_105540                         0.742892   0.595883                0.355403  0.452774  0.205004
# XRT_1_AutoML_20181212_105540                         0.742091   0.599346                0.356583  0.453117  0.205315
# DeepLearning_grid_1_AutoML_20181212_105540_model_2   0.741795   0.601497                0.368291  0.454904  0.206937
# XGBoost_grid_1_AutoML_20181212_105540_model_1        0.693554   0.620702                0.40588   0.465791  0.216961
# DeepLearning_1_AutoML_20181212_105540                0.69137    0.637954                0.409351  0.47178   0.222576
# DeepLearning_grid_1_AutoML_20181212_105540_model_1   0.690084   0.661794                0.418469  0.476635  0.227181
# GLM_grid_1_AutoML_20181212_105540_model_1            0.682648   0.63852                 0.397234  0.472683  0.223429
#
# [22 rows x 6 columns]

# The leader model is stored here
aml.leader

In [4]:
import h2o

In [5]:
from h2o.automl import H2OAutoML

In [6]:
h2o.init()

Checking whether there is an H2O instance running at http://localhost:54321..... not found.
Attempting to start a local H2O server...
; Java HotSpot(TM) 64-Bit Server VM (build 25.91-b14, mixed mode)
  Starting server from C:\AI\pythonProject\venv\Lib\site-packages\h2o\backend\bin\h2o.jar
  Ice root: C:\Users\ASIA\AppData\Local\Temp\tmpohna0itu
  JVM stdout: C:\Users\ASIA\AppData\Local\Temp\tmpohna0itu\h2o_ASIA_started_from_python.out
  JVM stderr: C:\Users\ASIA\AppData\Local\Temp\tmpohna0itu\h2o_ASIA_started_from_python.err
  Server is running at http://127.0.0.1:54321
Connecting to H2O server at http://127.0.0.1:54321 ... successful.


--------------------------  -----------------------------
H2O_cluster_uptime:         03 secs
H2O_cluster_timezone:       Asia/Seoul
H2O_data_parsing_timezone:  UTC
H2O_cluster_version:        3.40.0.4
H2O_cluster_version_age:    1 month and 30 days
H2O_cluster_name:           H2O_from_python_ASIA_imfjl4
H2O_cluster_total_nodes:    1
H2O_cluster_free_memory:    3.539 Gb
H2O_cluster_total_cores:    4
H2O_cluster_allowed_cores:  4
H2O_cluster_status:         locked, healthy
H2O_connection_url:         http://127.0.0.1:54321
H2O_connection_proxy:       {"http": null, "https": null}
H2O_internal_security:      False
Python_version:             3.8.8 final
--------------------------  -----------------------------

In [7]:
df = pd.concat([train,test], axis=0)

In [8]:
x= list(df[['장중_긍정비율','장중_부정비율','장중_긍부정비율','장외_긍정비율','장외_부정비율','장외_긍부정비율']])
y = '등락률'
# X_test = test[['장중_긍정비율','장중_부정비율','장중_긍부정비율','장외_긍정비율','장외_부정비율','장외_긍부정비율']]
# y_test= test['등락률']

In [9]:
h2o_train = h2o.H2OFrame(train)
h2o_test= h2o.H2OFrame(test)

Parse progress: |████████████████████████████████████████████████████████████████| (done) 100%
Parse progress: |████████████████████████████████████████████████████████████████| (done) 100%


In [10]:
# train[y] = train[y].asfactor()
# test[y] = test[y].asfactor()

In [11]:
aml = H2OAutoML(max_models=20, seed=1)
aml.train(x=x, y=y, training_frame=h2o_train)

AutoML progress: |
11:12:48.634: AutoML: XGBoost is not available; skipping it.

███████████████████████████████████████████████████████████████| (done) 100%


Model Details
=============
H2OStackedEnsembleEstimator : Stacked Ensemble
Model Key: StackedEnsemble_BestOfFamily_1_AutoML_1_20230628_111248


Model Summary for Stacked Ensemble: 
key                                        value
-----------------------------------------  ----------------
Stacking strategy                          cross_validation
Number of base models (used / total)       3/5
# GBM base models (used / total)           0/1
# GLM base models (used / total)           1/1
# DeepLearning base models (used / total)  1/1
# DRF base models (used / total)           1/2
Metalearner algorithm                      GLM
Metalearner fold assignment scheme         Random
Metalearner nfolds                         5
Metalearner fold_column
Custom metalearner hyperparameters         None

ModelMetricsRegressionGLM: stackedensemble
** Reported on train data. **

MSE: 0.924498303472019
RMSE: 0.9615083481031348
MAE: 0.716377493054376
RMSLE: NaN
Mean Residual Deviance: 0.924498303472019
R^2: 0.2414407206488608
Null degrees of freedom: 493
Residual degrees of freedom: 490
Null deviance: 602.0652233083678
Residual deviance: 456.7021619151774
AIC: 1373.1302635691748

ModelMetricsRegressionGLM: stackedensemble
** Reported on cross-validation data. **

MSE: 0.9665960130089483
RMSE: 0.9831561488435844
MAE: 0.7275956825587729
RMSLE: NaN
Mean Residual Deviance: 0.9665960130089483
R^2: 0.2068991665013441
Null degrees of freedom: 493
Residual degrees of freedom: 491
Null deviance: 603.0990031932869
Residual deviance: 477.49843042642044
AIC: 1393.1277965034285

Cross-Validation Metrics Summary: 
                        mean      sd         cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid
----------------------  --------  ---------  ------------  ------------  ------------  ------------  ------------
mae                     0.727823  0.0677023  0.754975      0.633067      0.737563      0.698236      0.815277
mean_residual_deviance  0.967173  0.0955956  0.953089      0.823402      0.969738      1.00226       1.08737
mse                     0.967173  0.0955956  0.953089      0.823402      0.969738      1.00226       1.08737
null_deviance           120.62    7.90075    115.012       119.683       129.475       111.21        127.719
r2                      0.205318  0.056682   0.177533      0.256875      0.275739      0.161805      0.154635
residual_deviance       95.2302   6.66233    94.3558       88.9274       93.0949       93.2106       106.562
rmse                    0.982467  0.0491463  0.976263      0.907415      0.984753      1.00113       1.04277
rmsle                   nan       0          nan           nan           nan           nan           nan

[tips]
Use `model.explain()` to inspect the model.
--
Use `h2o.display.toggle_user_tips()` to switch on/off this section.

<pre>
ModelMetricsRegressionGLM: stackedensemble
** Reported on train data. **

MSE: 0.924498303472019
RMSE: 0.9615083481031348
MAE: 0.716377493054376
RMSLE: NaN
Mean Residual Deviance: 0.924498303472019
R^2: 0.2414407206488608
Null degrees of freedom: 493
Residual degrees of freedom: 490
Null deviance: 602.0652233083678
Residual deviance: 456.7021619151774
AIC: 1373.1302635691748
ModelMetricsRegressionGLM: stackedensemble
** Reported on cross-validation data. **

MSE: 0.9665960130089483
RMSE: 0.9831561488435844
MAE: 0.7275956825587729
RMSLE: NaN
Mean Residual Deviance: 0.9665960130089483
R^2: 0.2068991665013441
Null degrees of freedom: 493
Residual degrees of freedom: 491
Null deviance: 603.0990031932869
Residual deviance: 477.49843042642044
AIC: 1393.1277965034285
